components/script/script_thread.rs

```diff
@@ -1643,43 +1631,52 @@ impl ScriptThread {
+        // Perform a microtask checkpoint as the specifications says that *update the rendering*
+        // should be run in a task and a microtask checkpoint is always done when running tasks.
+        self.perform_a_microtask_checkpoint(can_gc);
+
+        // If there are pending reflows, they were probably caused by the execution of
+        // the microtask checkpoint above and we should spin the event loop one more
+        // time to resolve them.
+        self.schedule_rendering_opportunity_if_necessary();
+    }
+
+    // If there are any pending reflows and we are not having rendering opportunities
+    // driven by the compositor, then schedule the next rendering opportunity.
+    //
+    // TODO: This is a workaround until rendering opportunities can be triggered from a
+    // timer in the script thread.
+    fn schedule_rendering_opportunity_if_necessary(&self) {
+        // If any Document has active animations of rAFs, then we should be receiving
+        // regular rendering opportunities from the compositor (or fake animation frame
+        // ticks). In this case, don't schedule an opportunity, just wait for the next
+        // one.
+        if self.documents.borrow().iter().any(|(_, document)| {
+            document.animations().running_animation_count() != 0 ||
+                document.has_active_request_animation_frame_callbacks()
+        }) {
+            return;
+        }
 
-        // Note: the pipeline should be a navigable with a rendering opportunity,
-        // and we should use this opportunity to queue one task for each navigable with
-        // an opportunity in this script-thread.
-        let Some(document) = self.documents.borrow().find_document(pipeline_id) else {
-            warn!("Trying to update the rendering for closed pipeline {pipeline_id}.");
+        let Some((_, document)) = self.documents.borrow().iter().find(|(_, document)| {
+            !document.window().reflows_suppressed() && document.needs_reflow().is_some()
+        }) else {
             return;
         };
-        let window = document.window();
-        let task_manager = window.task_manager();
+
+        let task_manager = document.window().task_manager();
         let rendering_task_source = task_manager.rendering_task_source();
         let canceller = task_manager.task_canceller(TaskSourceName::Rendering);
 
-        if !self
-            .update_the_rendering_task_queued_for_pipeline
-            .borrow_mut()
-            .insert(pipeline_id)
-        {
-            return;
-        }
-
         // Queues a task to update the rendering.
         // <https://html.spec.whatwg.org/multipage/#event-loop-processing-model:queue-a-global-task>
         //
         // Note: The specification says to queue a task using the navigable's active
         // window, but then updates the rendering for all documents.
         let _ = rendering_task_source.queue_with_canceller(
```

Comment:
I think we should add a `queue_unconditionally` method to the rendering task source, and use it here instead, because if the document becomes inactive and there are no other messages received, it may lead to livelock(although I'm not completely sure thanks to [the mechanism here](https://github.com/servo/servo/blob/84c59e09b679baadafc30145ac752e37844ac0b8/components/script/task_queue.rs#L100), but we may as not rely on it for liveness here since it's buried elsewhere). 

For an example of `queue_unconditionally` , see https://github.com/servo/servo/blob/84c59e09b679baadafc30145ac752e37844ac0b8/components/script/task_source/networking.rs#L42